fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/builder/FaceBuilder.java

```diff
@@ -0,0 +1,188 @@
+	 * @param x2 The X-coordinate of the second vertex (V).
+	 * @param y2 The Y-coordinate of the second vertex (V).
+	 * @param texture A key corresponding to the texture to be applied for this element face.
+	 * @param cullFace If specified, this face need not render if a block is directly adjacent to the given one.
+	 * @param rotation The rotation of this texture (quarter-turns only).
+	 */
+	public FaceBuilder(int x1, int y1, int x2, int y2, TextureKey texture, Face cullFace, VariantSettings.Rotation rotation) {
+		this(x1, y1, x2, y2, texture, cullFace, rotation, -1);
+	}
+
+	/**
+	 * Create a new face builder with a given UV, texture key and cull face.
+	 *
+	 * @param x1 The X-coordinate of the first vertex (U).
+	 * @param y1 The Y-coordinate of the first vertex (U).
+	 * @param x2 The X-coordinate of the second vertex (V).
+	 * @param y2 The Y-coordinate of the second vertex (V).
+	 * @param texture A key corresponding to the texture to be applied for this element face.
+	 * @param cullFace If specified, this face need not render if a block is directly adjacent to the given one.
+	 */
+	public FaceBuilder(int x1, int y1, int x2, int y2, TextureKey texture, Face cullFace) {
+		this(x1, y1, x2, y2, texture, cullFace, VariantSettings.Rotation.R0, -1);
+	}
+
+	/**
+	 * Create a new face builder with a given UV and texture key.
+	 *
+	 * @param x1 The X-coordinate of the first vertex (U).
+	 * @param y1 The Y-coordinate of the first vertex (U).
+	 * @param x2 The X-coordinate of the second vertex (V).
+	 * @param y2 The Y-coordinate of the second vertex (V).
+	 * @param texture A key corresponding to the texture to be applied for this element face.
+	 */
+	public FaceBuilder(int x1, int y1, int x2, int y2, TextureKey texture) {
+		this(x1, y1, x2, y2, texture, Face.NONE, VariantSettings.Rotation.R0, -1);
+	}
+
+	/**
+	 * Create a new face builder with a given texture key.
+	 * Defaults UV data to be based on the position of the element whose face is being built.
+	 *
+	 * @param texture A key corresponding to the texture to be applied for this element face.
+	 */
+	public FaceBuilder(TextureKey texture) {
+		this(new Vector4i(0), texture, Face.NONE, VariantSettings.Rotation.R0, -1);
+	}
+
+	public JsonObject build() {
+		var face = new JsonObject();
+
+		var uv = new JsonArray();
+		uv.add(this.uv.x());
+		uv.add(this.uv.y());
+		uv.add(this.uv.z());
+		uv.add(this.uv.w());
+		face.add("uv", uv);
+
+		face.addProperty("texture", "#" + texture.getName());
+
+		if (cullFace != Face.NONE) {
+			face.addProperty("cullface", cullFace.name().toLowerCase());
+		}
+
+		if (rotation != VariantSettings.Rotation.R0) {
+			face.addProperty("rotation", rotation.name().substring(1));
+		}
+
+		if (this.tintIndex != -1) {
+			face.addProperty("tintindex", tintIndex);
+		}
+
+		return face;
+	}
+
+	public enum Face {
+		DOWN, UP, NORTH, SOUTH, WEST, EAST,
+		/**
+		 * Fallback value; do not use directly.
+		 */
+		NONE
```

Comment:
That's weird. You should maybe use `null` as a marker instead.

Reply:
Though I should ask, why not use the `Direction` from MC?

Reply:
Good point. If I'm not going to use the `NONE` fallback to begin with then it might as well take a `Direction` instead.